File: server/internal/usecase/interfaces/project.go
package interfaces

import (
	"archive/zip"
	"context"
	"errors"
	"net/url"

	"github.com/reearth/reearthx/account/accountdomain/user"

	"github.com/reearth/reearth/server/internal/app/i18n/message/errmsg"
	"github.com/reearth/reearth/server/internal/usecase"
	"github.com/reearth/reearth/server/pkg/i18n/message"
	"github.com/reearth/reearth/server/pkg/id"
	"github.com/reearth/reearth/server/pkg/project"
	"github.com/reearth/reearth/server/pkg/verror"
	"github.com/reearth/reearth/server/pkg/visualizer"
	"github.com/reearth/reearthx/account/accountdomain"
	"github.com/reearth/reearthx/usecasex"
	"github.com/spf13/afero"
)

type CreateProjectParam struct {
	WorkspaceID     accountdomain.WorkspaceID
	Visualizer      visualizer.Visualizer
	Name            *string
	Description     *string
	CoreSupport     *bool
	Visibility      *string
	IsDeleted       *bool
	ImportStatus    project.ProjectImportStatus
	ImportResultLog *map[string]any
	ProjectAlias    *string

	// metadata
	Readme  *string
	License *string
	Topics  *[]string
}

type UpdateProjectParam struct {
	ID             id.ProjectID
	Name           *string
	Description    *string
	Archived       *bool
	ImageURL       *url.URL
	DeleteImageURL bool
	ImportStatus   project.ProjectImportStatus
	SceneID        *id.SceneID
	Starred        *bool
	Deleted        *bool
	Visibility     *string
	ProjectAlias   *string

	// publishment
	PublicTitle       *string
	PublicDescription *string
	PublicImage       *string
	PublicNoIndex     *bool
	DeletePublicImage bool
	IsBasicAuthActive *bool
	BasicAuthUsername *string
	BasicAuthPassword *string
	EnableGa          *bool
	TrackingID        *string
}

type CreateProjectMetadataParam struct {
	ProjectID   id.ProjectID
	WorkspaceID accountdomain.WorkspaceID
	Readme      *string
	License     *string
	Topics      *[]string
	StarCount   *int64
	StarredBy   *[]string
}

type UpdateProjectMetadataParam struct {
	ID        id.ProjectID
	Readme    *string
	License   *string
	Topics    *[]string
	StarCount *int64
	StarredBy *[]string
}

type PublishProjectParam struct {
	ID     id.ProjectID
	Alias  *string
	Status project.PublishmentStatus
}

var (
	ErrProjectAliasIsNotSet    error = errors.New("project alias is not set")
	ErrProjectAliasAlreadyUsed       = verror.NewVError(
		errmsg.ErrKeyUsecaseInterfaceProjectAliasAlreadyUsed,
		errmsg.ErrorMessages[errmsg.ErrKeyUsecaseInterfaceProjectAliasAlreadyUsed],
		message.MultiLocaleTemplateData(map[string]any{}),
		nil,
	)
)

type ProjectListParam struct {
	Limit  *int64
	Offset *int64
}

type Project interface {
	Fetch(context.Context, []id.ProjectID, *usecase.Operator) ([]*project.Project, error)
	FindByWorkspace(context.Context, accountdomain.WorkspaceID, *string, *project.SortType, *usecasex.Pagination, *usecase.Operator) ([]*project.Project, *usecasex.PageInfo, error)
	FindStarredByWorkspace(context.Context, accountdomain.WorkspaceID, *usecase.Operator) ([]*project.Project, error)
	FindDeletedByWorkspace(context.Context, accountdomain.WorkspaceID, *usecase.Operator) ([]*project.Project, error)

	FindActiveById(context.Context, id.ProjectID, *usecase.Operator) (*project.Project, error)
	FindActiveByAlias(context.Context, string, *usecase.Operator) (*project.Project, error)
	FindByProjectAlias(context.Context, string, *usecase.Operator) (*project.Project, error)

	FindVisibilityByUser(context.Context, *user.User, bool, *usecase.Operator, *string, *project.SortType, *usecasex.Pagination, *ProjectListParam) ([]*project.Project, *usecasex.PageInfo, error)
	FindVisibilityByWorkspace(context.Context, accountdomain.WorkspaceID, bool, *usecase.Operator, *string, *project.SortType, *usecasex.Pagination, *ProjectListParam) ([]*project.Project, *usecasex.PageInfo, error)
	FindAll(context.Context, *string, *project.SortType, *usecasex.Pagination, *ProjectListParam, *[]string, *string) ([]*project.Project, *usecasex.PageInfo, error)
	UpdateVisibility(context.Context, id.ProjectID, string, *usecase.Operator) (*project.Project, error)

	Create(context.Context, CreateProjectParam, *usecase.Operator) (*project.Project, error)
	Update(context.Context, UpdateProjectParam, *usecase.Operator) (*project.Project, error)
	Delete(context.Context, id.ProjectID, *usecase.Operator) error

	Publish(context.Context, PublishProjectParam, *usecase.Operator) (*project.Project, error)
	CheckProjectAlias(context.Context, string, accountdomain.WorkspaceID, *id.ProjectID) (bool, error)
	CheckSceneAlias(context.Context, string, *id.ProjectID) (bool, error)

	ExportProjectData(context.Context, id.ProjectID, *zip.Writer, *usecase.Operator) (*project.Project, error)
	ImportProjectData(context.Context, string, *string, *[]byte, *usecase.Operator) (*project.Project, error)
	UpdateImportStatus(context.Context, id.ProjectID, project.ProjectImportStatus, *map[string]any, *usecase.Operator) (*project.ProjectMetadata, error)
	SaveExportProjectZip(context.Context, *zip.Writer, afero.File, map[string]any, *project.Project) error
}

File: web/src/services/gql/queries/project.ts
import { gql } from "@reearth/services/gql/__gen__";

export const GET_PROJECT = gql(`
  query GetProject($projectId: ID!) {
    node(id: $projectId, type: PROJECT) {
      id
      ... on Project {
        ...ProjectFragment
        scene {
          id
          alias
        }
      }
    }
  }
`);

export const GET_PROJECTS = gql(`
  query GetProjects($workspaceId: ID!, $pagination: Pagination, $keyword: String, $sort: ProjectSort) {
    projects(workspaceId: $workspaceId, pagination: $pagination, keyword: $keyword, sort: $sort) {
      edges {
        node {
          id
          ...ProjectFragment
          scene {
            id
            alias
          }
        }
      }
      pageInfo {
        endCursor
        hasNextPage
        hasPreviousPage
        startCursor
      }
      totalCount
    }
  }
`);

export const CHECK_PROJECT_ALIAS = gql(`
  query CheckProjectAlias($alias: String!, $workspaceId: ID!, $projectId: ID) {
    checkProjectAlias(alias: $alias, workspaceId:$workspaceId, projectId: $projectId) {
      alias
      available
    }
  }
`);

export const CREATE_PROJECT = gql(`
  mutation CreateProject(
    $workspaceId: ID!
    $visualizer: Visualizer!
    $name: String!
    $description: String!
    $coreSupport: Boolean
    $visibility: String
    $projectAlias: String
    $readme: String
    $license: String
    $topics: String
  ) {
    createProject(
      input: {
        workspaceId: $workspaceId
        visualizer: $visualizer
        name: $name
        description: $description
        coreSupport: $coreSupport
        visibility: $visibility
	      projectAlias: $projectAlias
        readme: $readme
        license: $license
        topics: $topics
      }
    ) {
      project {
        id
        name
        description
        coreSupport
      }
    }
  }
`);

export const UPDATE_PROJECT = gql(`
  mutation UpdateProject(
    $projectId: ID!
    $name: String
    $description: String
    $imageUrl: URL
    $publicTitle: String
    $publicDescription: String
    $publicImage: String
    $deleteImageUrl: Boolean
    $deletePublicImage: Boolean
    $enableGa: Boolean
    $trackingId: String
    $starred:Boolean
    $deleted: Boolean
    $projectAlias: String
    $visibility: String
  ) {
    updateProject(
      input: {
        projectId: $projectId
        name: $name
        description: $description
        imageUrl: $imageUrl
        publicTitle: $publicTitle
        publicDescription: $publicDescription
        publicImage: $publicImage
        deleteImageUrl: $deleteImageUrl
        deletePublicImage: $deletePublicImage
        enableGa: $enableGa
        trackingId: $trackingId
        starred: $starred
        deleted: $deleted
        projectAlias: $projectAlias
        visibility: $visibility
      }
    ) {
      project {
        id
        ...ProjectFragment
      }
    }
  }

`);

export const UPDATE_PROJECT_BASIC_AUTH = gql(`
  mutation UpdateProjectBasicAuth(
    $projectId: ID!
    $isBasicAuthActive: Boolean
    $basicAuthUsername: String
    $basicAuthPassword: String
  ) {
    updateProject(
      input: {
        projectId: $projectId
        isBasicAuthActive: $isBasicAuthActive
        basicAuthUsername: $basicAuthUsername
        basicAuthPassword: $basicAuthPassword
      }
    ) {
      project {
        id
        name
        isBasicAuthActive
        basicAuthUsername
        basicAuthPassword
      }
    }
  }
`);

export const PUBLISH_PROJECT = gql(`
 mutation PublishProject($projectId: ID!, $alias: String, $status: PublishmentStatus!) {
  publishProject(input: { projectId: $projectId, alias: $alias, status: $status }) {
    project {
      id
      alias
      publishmentStatus
    }
  }
}
`);

export const ARCHIVE_PROJECT = gql(`
  mutation ArchiveProject($projectId: ID!, $archived: Boolean!) {
    updateProject(input: { projectId: $projectId, archived: $archived }) {
      project {
        id
        isArchived
      }
    }
  }
`);

export const DELETE_PROJECT = gql(`
  mutation DeleteProject($projectId: ID!) {
    deleteProject(input: { projectId: $projectId }) {
      projectId
    }
  }
`);

export const GET_STARRED_PROJECTS = gql(`
  query GetStarredProjects($workspaceId: ID!) {
    starredProjects(workspaceId: $workspaceId) {
				nodes {
					id
					name
					starred
          scene {
            id
          }
				}
				totalCount
			}
  }
`);

export const EXPORT_PROJECT = gql(`
  mutation ExportProject($projectId: ID!) {
    exportProject(input: { projectId: $projectId }) {
      projectDataPath
    }
  }
`);

export const GET_DELETED_PROJECTS = gql(`
  query GetDeletedProjects($workspaceId: ID!) {
    deletedProjects(workspaceId: $workspaceId) {
			nodes {
				id
				name
				isDeleted
        imageUrl
        visibility
        starred
			}
			totalCount
		}
  }
`);

export const UPDATE_PROJECT_METADATA = gql(`
  mutation UpdateProjectMetadata(
    $project: ID!
    $readme: String
    $license: String
    $topics: String
  ) {
    updateProjectMetadata(
      input: {
        project: $project
        readme: $readme
        license: $license
        topics: $topics
      }
    ) {
      metadata {
        id
      ...ProjectMetadataFragment
      }
    }
  }
`);
